---
title: "Ocular_snRNAseq_Photoreceptor_CAG_GRK1-GFP_integrated"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    number_sections: yes
    code_folding: hide
---

# Abstract
This is the note for this [experiment](https://sparktx.benchling.com/s/etr-B28Gaqz06jf0K5FaU4RQ?m=slm-6OXl3xxuormeqA9WWhSh)  
1	GTSAMPLE1488	RGC NEFH-GFP (peripheral)
2	GTSAMPLE1104	Photoreceptor GRK1-GFP
3	GTSAMPLE1487	Photoreceptor CAG-GFP
These samples were sequenced together, but the 1 sample is for other project.  

The second two samples are related. The purpose is to identify a rare photoreceptor subtype (S-cones) as well as  
assess GFP expression in our isolated nuclei driven by both the GRK1 promoter (sample 2) versus CAG (sample 3).  

[The plasmid annotation to get GFP info]:  
(https://sparktx.benchling.com/sparktx/f/lib_lGByfIlf-registry/seq_dqZSXQn3-spl_transgene_512/edit)  


```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      comment = "")
```

```{r}
library(biomaRt)
library(Seurat)
library(randomcoloR)
library(scCustomize)
library(sctransform)
library(glmGamPoi)
library(tidyverse)
library(corrplot)
library(reshape2)
library(future)
options(future.globals.maxSize = 800 * 1024^2) # tested, use 600M
library(openxlsx)
library(patchwork)
library(HGNChelper)
library(httpgd)
library(Scillus)
library(fgsea)
library(msigdbr)
library(qs)
library(knitr)
Sys.setenv(PROJ_LIB = "/home/ubuntu/anaconda3/envs/r4.2.1/share/proj")
Sys.setenv(GDAL_DATA = "/home/ubuntu/anaconda3/envs/r4.2.1/share/gdal")
source("/mnt/data/apps/sc-type/R/gene_sets_prepare.R")
source("/mnt/data/apps/sc-type/R/sctype_score_.R")
setwd('~/Ocular_snRNAseq_Photoreceptor_CAG_GRK1-GFP')
set.seed(123)
```


# Load and combine sample feature_bc_matrix from cellranger 
```{r, eval = FALSE}
rds_path <- list.files('./nf-core/output/cellranger/cellranger/mtx_conversions/',
    pattern = "GTSAMPLE1104-[1-8]_matrix\\.rds|GTSAMPLE1487-[1-8]_matrix\\.rds", full.names = TRUE)
rds_names <-  gsub("_matrix.rds$", "", basename(rds_path))
rds_list <- lapply(rds_path, readRDS)
names(rds_list) <- rds_names

# combine samples
seuratObj <- merge(rds_list[[1]], y = rds_list[2:16], add.cell.ids = names(rds_list), project = "Ocular_Photoreceptor_CAG_GRK1-GFP")
# add sample names
cells <- rownames(seuratObj@meta.data)

# extract sample ids
seuratObj$sample <- sub("_[^_]*$", "", cells)

# add groups
group <- ifelse(grepl("GTSAMPLE1104", seuratObj$sample), "GRK1", "CAG")
seuratObj <- AddMetaData(seuratObj, metadata = group, col.name = "group")

```

# QC and selecting cells for further analysis

## Add %mitochondrial counts and cell cycle scores 
```{r, eval = FALSE}

## mitochondrial genes:
mt_genes <- c("ND1", "ND2", "COX1", "COX2", "ATP8", "ATP6", "COX3", "ND3", "ND4L", "ND4", "ND5", "ND6", "CYTB")

seuratObj[["percent.mt"]] <- PercentageFeatureSet(seuratObj, features = mt_genes)

s.genes <- cc.genes.updated.2019$s.genes
g2m.genes <- cc.genes.updated.2019$g2m.genes

seuratObj <- CellCycleScoring(seuratObj, s.features = s.genes, g2m.features = g2m.genes)
```

## QC
- Filter cells by:  
1, umi/nCount_RNA, low-quality cells (very low UMI counts) or doublets (very high UMI counts)  
2, gene/nFeature_RNA, remove potential low-quality cells and multiplets as above  
3, percent.mt, note that this step directly removes cells with high mt, the regress.out function later   
removes variations from mt and cell-cyle gene expression.   
4, our cutoff is chosen by visulizing data in Loupe browser    
5
== Load filtered Seurat obj instead of re-run the following steps 
```{r}
seuratObj <- qread("ocular_obj.qs")
# seuratObj <- readRDS("ocular_obj.rds")
```

- The object after pre-process
```{r}
seuratObj
table(seuratObj[[]]$sample)
head(seuratObj, 5)
```

### QC stats
```{r}
summary(seuratObj$nCount_RNA)
summary(seuratObj$nFeature_RNA)
summary(seuratObj$percent.mt)
```

UMI range is (`r min(seuratObj$nCount_RNA)`, `r max(seuratObj$nCount_RNA)`)
nfeatue range is (`r min(seuratObj$nFeature_RNA)`, `r max(seuratObj$nFeature_RNA)`)

Retinal ganglion cells are larger, results in a greater amount of total RNA per cell. 
Skip the nCount_RNA and nFeature_RNA filter as the gRed team did in their 2023 Cell Genomics paper.

Only remove cells with >5% mitochondrial expressions: `r length(seuratObj$percent.mt[seuratObj$percent.mt > 5])`

```{r, eval = FALSE}
seuratObj <- subset(seuratObj, subset = percent.mt < 5 &
    nFeature_RNA > 100 &
    nCount_RNA > 200)
```

### QC metrics as a violin plot 
- by initial clusters
```{r, fig.height=10, fig.width=12}
VlnPlot(seuratObj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), pt.size = 0, ncol = 3)
```

- by samples
```{r, fig.height=6, fig.width=12}
VlnPlot(seuratObj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
    group.by = 'sample',
    pt.size = 0, ncol = 1)
```


## SCTransform normalization, PCA and Clustering 
```{r, eval = FALSE}
# regressing out unwanted variation (mitochondrial, cell cyle, etc.)
seuratObj <- SCTransform(seuratObj, method = "glmGamPoi", vst.flavor = "v2",
        vars.to.regress = c("percent.mt", "S.Score", "G2M.Score"), verbose = TRUE)

seuratObj <- RunPCA(seuratObj, npcs = 40)
# ElbowPlot(seuratObj, ndims = 40)
seuratObj <- RunUMAP(seuratObj, reduction = "pca", dims = 1:30)
seuratObj <- FindNeighbors(seuratObj, reduction = "pca", dims = 1:30)
seuratObj <- FindClusters(seuratObj, resolution = 0.1)
table(seuratObj$seurat_clusters)
```

== Save processed object
```{r, eval = FALSE}
qsave(seuratObj, file = "ocular_obj.qs")
# saveRDS(seuratObj, file = "ocular_obj.rds")
```

## Visualize initial clusters
```{r, fig.height=7, fig.width=14}
p1 <- DimPlot(seuratObj, reduction = "umap", group.by = "sample") + theme(legend.position = "top")
p2 <- DimPlot(seuratObj, reduction = "umap", label = TRUE, repel = TRUE) + theme(legend.position = "top")
p1 + p2
```


```{r, fig.height=10, fig.width=14}
DimPlot(seuratObj, reduction = "umap", label = TRUE, split.by = "sample", ncol = 4) + theme(legend.position = "top")
```

## Number of cells in each cluster
```{r}
plot_clusters <- function(srat) {
    count_table <- table(srat@meta.data$seurat_clusters, srat@meta.data$sample)
    count_mtx <- as.data.frame.matrix(count_table)
    count_mtx$clusters <- rownames(count_mtx)
    melt_mtx <- melt(count_mtx)

    celltype_size <- aggregate(value ~ clusters, data = melt_mtx, FUN = sum)
    celltype_size <- celltype_size %>% dplyr::arrange(value)
    celltype_size$clusters <- factor(celltype_size$clusters, levels = celltype_size$clusters)
    melt_mtx$clusters <- factor(melt_mtx$clusters, levels = celltype_size$clusters)
    colnames(melt_mtx)[2] <- "sample"

    p1 <- ggplot(celltype_size, aes(y = clusters, x = value)) +
    geom_bar(position = "dodge", stat = "identity", fill = "grey60") +
    theme_bw() + scale_x_log10() + xlab("Cells per cluster, log10 scale") + ylab("")

    p2 <- ggplot(melt_mtx, aes(x = clusters, y = value, fill = sample)) +
        geom_bar(position = "fill", stat = "identity") + theme_bw() + coord_flip() +
        scale_fill_manual(values=as.vector(randomColor(16))) +
        ylab("Fraction of cells in each sample") + xlab("cluster number") + theme(legend.position = "top")

    p2 + p1 + plot_layout(widths = c(3, 1))
}
```
```{r fig.height=8, fig.width=14}
plot_clusters(seuratObj)
```


# Data Integration  
To identify cell types present in all samples/conditions/modalities within our dataset,  
and would like to observe a representation of cells from both samples/conditions/modalities in every cluster.  
Reference: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/06_integration.md  
- Execute the normalization, variance estimation, and identification of the most variable genes for each sample   
- Perform integration of cells across conditions using the most variant genes to identify cells most similar to each other  

== Load object list `seuratObj.list`
```{r, eval = FALSE}
seuratObj.list <- qread("ocular_obj.list.qs")
# seuratObj.list <- readRDS("ocular_obj.list.rds")
```

== Load objects `ocular_obj.integrated`
```{r}
seuratObj.integrated <- qread("ocular_obj.integrated.qs")
# seuratObj.integrated <- readRDS("ocular_obj.integrated.rds")
table(seuratObj.integrated$seurat_clusters)
```

## Split object and SCTransform normalization  
split the dataset into a list of two seurat objects (GLA_KO and WT), 
normalize and identify variable features for each dataset independently.  
```{r, eval = FALSE}
seuratObj.list <- SplitObject(seuratObj, split.by = "sample")

for (i in 1:length(seuratObj.list)) {
    seuratObj.list[[i]] <- SCTransform(seuratObj.list[[i]], method = "glmGamPoi", vst.flavor = "v2",
        vars.to.regress = c("percent.mt", "S.Score", "G2M.Score"), verbose = TRUE, return.only.var.genes = TRUE) %>%
    RunPCA(npcs = 30)
}
```

## Select features and prepare integration 
Select features that are repeatedly variable across datasets for integration, 
and run `PrepSCTIntegration`, which ensures that all necessary Pearson residuals have been calculated.  
```{r, eval = FALSE}
## "ENSMFAG00000043960" is OPN1LW,  "ENSMFAG00000036911" is RBPMS, they don't not have gene symbols in MacFas6.0 gtf
selected_markers <- c("GRK1", "GRK7", "ABCA4", "RHO", "NRL", "GNAT1", "ARR3", "GNAT2", 
    "OPN1SW", "ENSMFAG00000043960", "ONECUT1", "VSX2", "GAD1", "ENSMFAG00000036911", "RLBP1", "BEST1")
seuratObj.features <- SelectIntegrationFeatures(object.list = seuratObj.list, nfeatures = 3000)
seuratObj.features <- unique(c(seuratObj.features, selected_markers))
seuratObj.list <- PrepSCTIntegration(object.list = seuratObj.list, anchor.features = seuratObj.features)
```

== Save object `seuratObj.list`
```{r, eval = FALSE}
qsave(seuratObj.list, file = "ocular_obj.list.qs")
# saveRDS(seuratObj.list, file = "ocular_obj.list.rds")
```

## Identify anchors and integrate the datasets
### using the samples with high RGCs as references (periphery samples here)
When using a set of specified references, anchors are first found between each query and each reference. 
The references are then integrated through pairwise integration. Each query is then mapped to the integrated reference.
Make sure to set normalization.method = **"SCT"**, and reduction = **"rpca"** for reciprocal PCA  
```{r, eval = FALSE}
anchors <- FindIntegrationAnchors(object.list = seuratObj.list, normalization.method = "SCT",
    reference = seq(9,16),
    anchor.features = seuratObj.features, reduction = "rpca")
seuratObj.integrated <- IntegrateData(anchorset = anchors, normalization.method = "SCT")
```


# Cluster on the integrated dataset
## PCA and Clustering  
```{r, eval = FALSE}
DefaultAssay(seuratObj.integrated) <- "integrated"
seuratObj.integrated <- RunPCA(seuratObj.integrated, npcs = 30)
# ElbowPlot(seuratObj.integrated, ndims = 30)

seuratObj.integrated <- RunUMAP(seuratObj.integrated, reduction = "pca", dims = 1:30)
seuratObj.integrated <- FindNeighbors(seuratObj.integrated, reduction = "pca", dims = 1:30)
# try 0.3-0.5, give 26, 28, 30 clusters
seuratObj.integrated <- FindClusters(seuratObj.integrated, resolution = 0.1)
table(seuratObj.integrated$seurat_clusters)
```

== Save integrated object
```{r, eval = FALSE}
qsave(seuratObj.integrated, file = "ocular_obj.integrated.qs")
# saveRDS(seuratObj.integrated, file = "ocular_obj.integrated.rds")
```

## Plots after integration   
```{r, fig.height=7, fig.width=14}
p1 <- DimPlot(seuratObj.integrated, reduction = "umap", group.by = "sample") + theme(legend.position = "top")
p2 <- DimPlot(seuratObj.integrated, reduction = "umap", label = TRUE, repel = TRUE) + NoLegend()
p1 + p2
```

```{r, fig.height=18, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = FALSE, split.by = "sample", ncol = 4) + theme(legend.position = "top")
```

```{r, fig.height=10, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = FALSE, split.by = "group", ncol = 2) + theme(legend.position = "top")
```

## Number of cells in each cluster
```{r}
plot_integrated_clusters <- function(srat) { 
    count_table <- table(srat@meta.data$seurat_clusters, srat@meta.data$sample)
    count_mtx <- as.data.frame.matrix(count_table)
    count_mtx$clusters <- rownames(count_mtx)
    melt_mtx <- melt(count_mtx)

    celltype_size <- aggregate(value ~ clusters, data = melt_mtx, FUN = sum)
    celltype_size <- celltype_size %>% dplyr::arrange(value)
    celltype_size$clusters <- factor(celltype_size$clusters, levels = celltype_size$clusters)
    melt_mtx$clusters <- factor(melt_mtx$clusters, levels = celltype_size$clusters)
    colnames(melt_mtx)[2] <- "sample"

    p1 <- ggplot(celltype_size, aes(y = clusters, x = value)) +
    geom_bar(position = "dodge", stat = "identity", fill = "grey60") +
    theme_bw() + scale_x_log10() + xlab("Cells per cluster, log10 scale") + ylab("")

    p2 <- ggplot(melt_mtx, aes(x = clusters, y = value, fill = sample)) +
        geom_bar(position = "fill", stat = "identity") + theme_bw() + coord_flip() +
        scale_fill_manual(values=as.vector(distinctColorPalette(20))) +
        ylab("Fraction of cells in each sample") + xlab("cluster number") + theme(legend.position = "top")

    p2 + p1 + plot_layout(widths = c(3, 1))
}
```
```{r fig.height=8, fig.width=14}
plot_integrated_clusters(seuratObj.integrated)
```

# Cell type annotation
## Apply marker based tool `sc-type`
```{r}
sctype <- function(obj) {
    # prepare marker genes from the database, category: Immune system, Liver, Pancreas, Kidney, Eye, Brain
    gs_list <- gene_sets_prepare("/mnt/data/apps/sc-type/ScTypeDB_short_NHPeye.xlsx", "NHP_Eye")

    # use a customized marker list
    # gs_list <- gene_sets_prepare("./ocular_customized_markers.xlsx", "Eye")
    # get cell-type by cell matrix
    # NOTE: scRNAseqData parameter should correspond to your input scRNA-seq matrix.
    # In case Seurat is used, it is either pbmc[["RNA"]]@scale.data (default),
    # pbmc[["SCT"]]@scale.data, in case sctransform is used for normalization,
    # or pbmc[["integrated"]]@scale.data, in case a joint analysis of multiple single-cell datasets is performed.
    es.max <- sctype_score(scRNAseqData = obj[['SCT']]@scale.data, scaled = TRUE,
                        gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)

    # merge by cluster
    metadata <- obj@meta.data
    clusters <- unique(obj@meta.data$seurat_clusters)
    cL_resutls <- do.call("rbind",
        lapply(
            clusters,
            function(cl) {
                es.max.cl <- sort(rowSums(es.max[, rownames(metadata[metadata$seurat_clusters == cl, ])]),
                    decreasing = TRUE)
                cell_scores <- data.frame(cluster = cl,
                        type = names(es.max.cl),
                        scores = es.max.cl,
                        ncells = sum(metadata$seurat_clusters == cl))
                head(cell_scores, 10)
            }
        )
    )
    sctype_scores <- cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)

    # set low-confident (low ScType score) clusters to "unknown"
    sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] <- "Unknown"
    print(data.frame(sctype_scores))

    # add cell type annotations to obj
    obj@meta.data$celltypes <- ""
    for (j in unique(sctype_scores$cluster)){
    cl_type <- sctype_scores[sctype_scores$cluster == j, ]
    obj@meta.data$celltypes[obj@meta.data$seurat_clusters == j] <-
        as.character(cl_type$type[1])
    }
    return(obj)
}
```

```{r}
seuratObj.integrated <- sctype(seuratObj.integrated)
```


## Show cell types on UMAP
```{r, fig.height=14, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = TRUE, repel = TRUE,
    group.by = 'celltypes') +
    theme(legend.position = "top")
table(seuratObj.integrated$celltypes)
```

```{r, fig.height=14, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = TRUE, repel = TRUE, split.by = 'sample', ncol = 4,
    group.by = 'celltypes') +
    theme(legend.position = "top")
```

```{r, fig.height=10, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = TRUE, repel = TRUE, split.by = 'group', ncol = 2,
    group.by = 'celltypes') +
    theme(legend.position = "top")
```


## Number of cells in each cell type
```{r}
plot_celltypes <- function(srat) {
    count_table <- table(srat@meta.data$celltypes, srat@meta.data$sample)
    count_mtx <- as.data.frame.matrix(count_table)
    count_mtx$celltypes <- rownames(count_mtx)
    melt_mtx <- melt(count_mtx)

    celltype_size <- aggregate(value ~ celltypes, data = melt_mtx, FUN = sum)
    celltype_size <- celltype_size %>% dplyr::arrange(value)
    celltype_size$celltypes <- factor(celltype_size$celltypes, levels = celltype_size$celltypes)
    melt_mtx$celltypes <- factor(melt_mtx$celltypes, levels = celltype_size$celltypes)
    colnames(melt_mtx)[2] <- "sample"

    p1 <- ggplot(celltype_size, aes(y = celltypes, x = value)) +
    geom_bar(position = "dodge", stat = "identity", fill = "grey60") +
    theme_bw() + scale_x_log10() + xlab("Cells per celltype, log10 scale") + ylab("")

    p2 <- ggplot(melt_mtx, aes(x = celltypes, y = value, fill = sample)) +
        geom_bar(position = "fill", stat = "identity") + theme_bw() + coord_flip() +
        scale_fill_manual(values=as.vector(randomColor(16))) +
        ylab("Fraction of cells in each sample") + xlab("celltype number") + theme(legend.position = "top")

    p2 + p1 + plot_layout(widths = c(3, 1))
}
```

```{r fig.height=6, fig.width=14}
plot_celltypes(seuratObj.integrated)
```



# Cluster markers analysis
== Load all cluster markers obj
```{r}
all.markers <- readRDS("all.cluster_markers.rds")
de_genes <- readRDS("de_genes_rod_vs_cone.rds")
```

## Find all markers
```{r, eval = FALSE}
Idents(seuratObj.integrated) <- 'celltypes'
seuratObj.integrated <- PrepSCTFindMarkers(seuratObj.integrated, assay = "SCT", verbose = TRUE)
plan("multicore", workers = 12)
all.markers <- FindAllMarkers(seuratObj.integrated, assay = "SCT", verbose = TRUE)
plan(sequential)
```

## Save marker genes
```{r, eval = FALSE}
all.markers <- all.markers %>% relocate(gene)
write_tsv(all.markers, "all.cluster_markers.tsv")
top20 <- all.markers %>%
    group_by(cluster) %>%
    top_n(n = 25, wt = avg_log2FC)
write_tsv(top20, "top25.cluster_markers.tsv")
saveRDS(all.markers, file = "all.cluster_markers.rds")
```

## Heatmap with top genes {.tabset}

### Markers for each Cluster
- showing expression in 10% random cells for each cluster  
```{r,  eval = FALSE, fig.height=20, fig.width=12}
Idents(seuratObj.integrated) <- "seurat_clusters"
DefaultAssay(seuratObj.integrated) <- "SCT"
top10 <- all.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC)

DoHeatmap(subset(seuratObj.integrated, downsample = 100), features = top10$gene, size = 3) +
    theme(legend.position = 'none', plot.margin = margin(t = 20))
```

# RETINAL GANGLION CELLS (RGCs) ==========

```{r, fig.height=14, fig.width=14}
DefaultAssay(seuratObj.integrated) <- "SCT"
table(Idents(seuratObj.integrated))
Idents(seuratObj.integrated) <- "celltypes"
# hgd()
# hgd_view()

# RGCs are the projection neurons of the retina
# markers of RGCs are:

VlnPlot(object = seuratObj.integrated, features = c("ENSMFAG00000036911", "SLC17A6", "SNCG", "SYT2"), pt.size = 0.01, combine = TRUE)

# RBPMS, which is not correctly annotated in macaque reference genome 
# transcript ID for this is "ENSMFAG00000036911"

```

## plot % RGCs for each sample/tissue/source
```{r}
Idents(seuratObj.integrated) <- "celltypes"
cell_type_to_plot <- "Retinal ganglion cells"
percentage_data <- seuratObj.integrated@meta.data %>%
    group_by(sample) %>%
    summarise(Percentage = sum(celltypes == cell_type_to_plot) / n()) %>%
    ungroup()

# pdf("Percent_RGC_per_sample.pdf", 8,6)
ggplot(percentage_data, aes(x = sample, y = Percentage, fill = sample)) +
    geom_bar(stat = "identity") +
    labs(title = paste("Percentage of", cell_type_to_plot), fill = "sample") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1), plot.title = element_text(hjust = 0.5)) +
    scale_y_continuous(labels = scales::percent) +
    theme(plot.margin = margin(5, 5, 5, 20, "mm")) 
# dev.off()
```

# Photoreceptors =====

**GRK1 stats**
```{r}
kable(all.markers[all.markers$gene == "GRK1" & all.markers$cluster == 'Rod photoreceptors', ])
```

## plot % Rods and Cones for each sample/tissue/source
```{r}
Idents(seuratObj.integrated) <- "celltypes"
cell_type_to_plot <- c("Rod photoreceptors", 'Cone photoreceptors')
percentage_data <- seuratObj.integrated@meta.data %>%
    count(sample, celltypes) %>%
    # Calculate the percentage
    group_by(sample) %>%
    mutate(Percentage = n / sum(n)) %>%
    filter(celltypes %in% cell_type_to_plot) %>%
    ungroup()

# pdf("Percent_RGC_per_sample.pdf", 8,6)
ggplot(percentage_data, aes(x = sample, y = Percentage, fill = sample)) +
    geom_bar(stat = "identity") +
    labs(title = paste("Percentage of", cell_type_to_plot), fill = "sample") +
    theme_minimal() +
    facet_wrap(~celltypes) + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1), plot.title = element_text(hjust = 0.5)) +
    scale_y_continuous(labels = scales::percent) +
    theme(plot.margin = margin(5, 5, 5, 20, "mm")) 
# dev.off()
```

## Check Photoreceptors markers
### Violin plots
```{r, fig.height=16, fig.width=16}
Idents(seuratObj.integrated) <- "celltypes"
DefaultAssay(seuratObj.integrated) <- "SCT"
rod_markers <- c("GNAT1","PDE6B", "RHO", "ABCA4")
cone_markers <- c("GRK7", "PDE6C", "ARR3")
VlnPlot(object = seuratObj.integrated, features = c("GRK1", rod_markers, cone_markers), pt.size = 0.01, combine = TRUE)
```

### UMAPs
```{r, fig.height=8, fig.width=8}
DimPlot(seuratObj.integrated, reduction = "umap", label = TRUE, repel = TRUE, group.by = 'celltypes') + theme(legend.position = "top")
```

```{r, fig.height=18, fig.width=18}
FeaturePlot(object = seuratObj.integrated, features = c("GRK1", rod_markers, cone_markers), label =TRUE, combine = TRUE)
```

### Dot plot for markers in each cell type
- showing top markers with customized markers
```{r, fig.height=8, fig.width=10}
Idents(seuratObj.integrated) <- "celltypes"
DefaultAssay(seuratObj.integrated) <- "SCT"
top_genes <- all.markers %>%
    filter(cluster %in% c('Rod photoreceptors', 'Cone photoreceptors')) %>%
    group_by(cluster) %>%
    slice_max(order_by = avg_log2FC, n = 10)

# rod_markers <- c("GNAT1","PDE6B","SAMD11",'PITPNM3',"PRPH2")
# cone_markers <- c("PDE6H", "PDE6C", "GNGT2", "ARR3","OPN1SW")

# Specify any additional genes you want to label
additional_genes <- c("GRK1", "RHO", "ABCA4", "ARR3")

# Combine top genes and additional genes, ensuring uniqueness
label_genes <- unique(c(top_genes$gene, additional_genes))
# label_genes <- unique(c(rod_markers, cone_markers, additional_genes))

DotPlot(seuratObj.integrated, features = label_genes, cols = c("white", 'blue'), dot.scale = 8) +
    RotatedAxis()
```


### Differential expression between Rods and Cones
```{r, eval = FALSE}

seuratObj.integrated <- PrepSCTFindMarkers(seuratObj.integrated, assay = "SCT", verbose = TRUE)

plan("multicore", workers = 12)
## set minimum cutoffs to include GRK1 if it is not significantly different among the groups
de_genes <- FindMarkers(seuratObj.integrated, ident.1 = 'Rod photoreceptors', ident.2 = 'Cone photoreceptors', min.pct = 0.01, 
    logfc.threshold = 0,
    p_val = 1,
    assay = "SCT", verbose = TRUE)
plan(sequential)

de_genes <- rownames_to_column(de_genes, var = "gene")
write_tsv(de_genes, "de_genes_rod_vs_cone.tsv")
saveRDS(de_genes, file = "de_genes_rod_vs_cone.rds")
```

```{r}
up_gene <- de_genes %>%
    slice_max(order_by = avg_log2FC, n = 10)

down_gene <- de_genes %>%
    slice_min(order_by = avg_log2FC, n = 10)
```

**Significantly up-regulated genes in Rods(top50)**
```{r}
kable(up_gene)
```
**Significantly up-regulated genes in Cones(top50)**
```{r}
kable(down_gene)
```

**GRK1 stats**
```{r}
kable(de_genes[de_genes$gene == "GRK1", ])
```


#### volcano plot
```{r}
# Add a column for -log10(p-value)
de_genes$p_val_adj_log10 <- -log10(de_genes$p_val_adj)

# Determine the top 10 genes based on adjusted P-value
top_genes <- de_genes %>%
    slice_max(order_by = avg_log2FC, n = 5)

bottom_genes <- de_genes %>%
    slice_min(order_by = avg_log2FC, n = 5)

# Specify any additional genes you want to label
additional_genes <- c("GRK1", "RHO", "ABCA4", "ARR3")

# Combine top genes and additional genes, ensuring uniqueness
label_genes <- unique(c(top_genes$gene, bottom_genes$gene, additional_genes))

# Create the volcano plot
volcano_plot <- ggplot(de_genes, aes(x = avg_log2FC, y = p_val_adj_log10, label = gene)) +
    geom_point(alpha = 0.5) + 
    labs(title = "Volcano Plot of Differential Expression (Rod vs Cone)",
        x = "Average Log2 Fold Change",
        y = "-Log10 Adjusted P-Value") +
    geom_text(data = filter(de_genes, gene %in% label_genes), vjust = 1.5, hjust = 0.5,  color = "red", check_overlap = TRUE) +
    theme_minimal() +
    theme(plot.title = element_text(size = 16, hjust = 0.5, margin = margin(b = 20)))

# Print the plot
print(volcano_plot)

```


# Plots to report
## rename cell types
```{r}
rename_map <- c("Rod photoreceptors" = "Rod",
                "Cone photoreceptors" = "Cone",
                "Amacrine cells" = "AC",
                "Retinal ganglion cells" = "RGC",
                "Muller glia" = "MC",
                "Bipolar cells" = "BC")

seuratObj.integrated$celltypes <- ifelse(seuratObj.integrated$celltypes %in% names(rename_map), rename_map[seuratObj.integrated$celltypes], seuratObj.integrated$celltypes)
Idents(seuratObj.integrated) <- "celltypes"

celltypes <- c("Rod", "Cone", "BC", "AC", "RGC", "MC")
sub_retina <- subset(seuratObj.integrated, idents = as.vector(celltypes))
table(sub_retina$celltypes)
```

## DotPlot for selected markers
```{r}
DefaultAssay(sub_retina) <- "SCT"
Idents(sub_retina) <- "celltypes"

celltypes <- c("Rod", "Cone", "BC", "AC", "RGC", "MC")
sub_retina$celltypes <- factor(sub_retina$celltypes, levels = celltypes)

markers <- c("GRK1", "GRK7", "ABCA4", "RHO", "NRL", "GNAT1", "ARR3", "GNAT2", 
    "OPN1SW", "ENSMFAG00000043960", "ONECUT1", "VSX2", "GAD1", "ENSMFAG00000036911", "RLBP1", "BEST1")

DotPlot(sub_retina, features = rev(markers), cols = c("white", 'blue'),  group.by = "celltypes", dot.scale = 10, scale = TRUE) + 
    RotatedAxis() + coord_flip() +
    scale_x_discrete(labels = c("ENSMFAG00000043960" = "OPN1LW", "ENSMFAG00000036911" = "RBPMS"))
```

## Heatmap for selected markers
** randomly sample 100 cells for each cell type **
```{r}
DoHeatmap(subset(sub_retina, downsample = 100), features = markers, group.by = "celltypes", size = 3, slot = "scale.data") + 
    theme(plot.margin = margin(t = 20)) + scale_y_discrete(labels = c("ENSMFAG00000043960" = "OPN1LW", "ENSMFAG00000036911" = "RBPMS"))

```

** sample 5% of cells for each cell type **
```{r}
sampled_sub_seurat <- qread('sampled_sub_seurat.qs')
table(sampled_sub_seurat$celltypes)
```

```{r, eval = FALSE}
seurat_list <- SplitObject(sub_retina, split.by = "celltypes")
# Sample 10% of cells from each cell type
sampled_list <- lapply(seurat_list, function(x) {
    # Calculate number of cells to sample
    num_to_sample <- ceiling(ncol(x) * 0.05)
    # Randomly sample cells
    sampled_cells <- sample(colnames(x), num_to_sample)
    # Subset the Seurat object for sampled cells
    subset(x, cells = sampled_cells)
})
# Merge the sampled objects back into a single Seurat object
sampled_sub_seurat <- Reduce(function(x, y) merge(x, y), sampled_list)
table(sampled_sub_seurat$celltypes)
```

```{r, eval = FALSE}
qsave(sampled_sub_seurat, file = 'sampled_sub_seurat.qs')
```


```{r}
Idents(sampled_sub_seurat) <- "celltypes"
celltypes <- c("Rod", "Cone", "BC", "AC", "RGC", "MC")
sampled_sub_seurat$celltypes <- factor(sampled_sub_seurat$celltypes, levels = celltypes)
levels(sampled_sub_seurat$celltypes)

DoHeatmap(sampled_sub_seurat, features = markers, group.by = "celltypes", size = 3,  slot = "scale.data") +
    theme(plot.margin = margin(t = 20)) + scale_y_discrete(labels = c("ENSMFAG00000043960" = "OPN1LW", "ENSMFAG00000036911" = "RBPMS"))

```