---
title: "Ocular_snRNAseq_macula_periphery_integration"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    number_sections: yes
    code_folding: hide
---

# Abstract
Integrate the three data sets from control eyes with macula and peripheral samples?   
It's important to keep tissue type in meta data as I'd like to see if in overall we saw higher %RGCs in macula samples.  
To confirm the datasets, they are 240119_VH00163_79_AACCFW3HV, 240124_VH00163_80_AACCNCGHV, and 240201_VH00163_82_AACCNHKHV.  

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      comment = "")
```

```{r}
library(biomaRt)
library(Seurat)
library(randomcoloR)
library(scCustomize)
library(sctransform)
library(glmGamPoi)
library(tidyverse)
library(corrplot)
library(reshape2)
library(future)
options(future.globals.maxSize = 800 * 1024^2) # tested, use 600M
library(openxlsx)
library(patchwork)
library(HGNChelper)
library(httpgd)
library(Scillus)
library(fgsea)
library(msigdbr)
library(qs)
Sys.setenv(PROJ_LIB = "/home/ubuntu/anaconda3/envs/r4.2.1/share/proj")
Sys.setenv(GDAL_DATA = "/home/ubuntu/anaconda3/envs/r4.2.1/share/gdal")
source("/mnt/data/apps/sc-type/R/gene_sets_prepare.R")
source("/mnt/data/apps/sc-type/R/sctype_score_.R")
setwd('~/Ocular_snRNAseq_macula_periphery_integration')
set.seed(123)
```


# Load and combine sample feature_bc_matrix from cellranger 
```{r, eval = FALSE}
# List all the files with ".rds$" and "_S" in their names
rds_path1 <- list.files('~/Ocular_snRNAseq_macula_periphery/nf-core/output/cellranger/cellranger/mtx_conversions/',
    pattern = "_matrix\\.rds", full.names = TRUE)
rds_names1 <-  gsub("_matrix.rds$", "", basename(rds_path1))
rds_list1 <- lapply(rds_path1, readRDS)
names(rds_list1) <- rds_names1

# dataset2
rds_path2 <- list.files('~/Ocular_snRNAseq_macula_periphery_Wunsorted/nf-core/output/cellranger/cellranger/mtx_conversions/',
    pattern = "_matrix\\.rds", full.names = TRUE)
# remove the first two bad samples
rds_path2 <- rds_path2[3:8]
rds_names2 <-  gsub("_matrix.rds$", "", basename(rds_path2))
rds_list2 <- lapply(rds_path2, readRDS)
names(rds_list2) <- rds_names2

# dataset3
rds_path3 <- list.files('~/Ocular_snRNAseq_macula_comp_dissections/nf-core/output/cellranger/cellranger/mtx_conversions/',
    pattern = "_matrix\\.rds", full.names = TRUE)
rds_names3 <-  gsub("_matrix.rds$", "", basename(rds_path3))
rds_list3 <- lapply(rds_path3, readRDS)
names(rds_list3) <- rds_names3

rds_list <- c(rds_list1, rds_list2, rds_list3)


# combine samples
seuratObj <- merge(rds_list[[1]], y = rds_list[2:20], add.cell.ids = names(rds_list), project = "ocular_macula_RGC_integration")
# add sample names
cells <- rownames(seuratObj@meta.data)

# extract sample ids
seuratObj$sample <- sub("_[^_]*$", "", cells)

# add groups
group <- ifelse(grepl("GTSAMPLE1006|GTSAMPLE1007", seuratObj$sample), "macula_periphery_set1", ifelse(grepl("GTSAMPLE1044|GTSAMPLE1045", seuratObj$sample), "macula_sort_unsort_set2", "macula_comp_dissection_set3"))
seuratObj <- AddMetaData(seuratObj, metadata = group, col.name = "group")

sample = unique(seuratObj$sample)

#  [1] "GTSAMPLE1006_1"          "GTSAMPLE1006_2"          "GTSAMPLE1006_3"          "GTSAMPLE1007_1"          "GTSAMPLE1007_2"         
#  [6] "GTSAMPLE1007_3"          "GTSAMPLE1044-sorted-3"   "GTSAMPLE1044-unsorted-1" "GTSAMPLE1044-unsorted-2" "GTSAMPLE1045-1"         
# [11] "GTSAMPLE1045-2"          "GTSAMPLE1045-3"          "GTSAMPLE1051-1"          "GTSAMPLE1051-2"          "GTSAMPLE1051-3"         
# [16] "GTSAMPLE1051-4"          "GTSAMPLE1052-1"          "GTSAMPLE1052-2"          "GTSAMPLE1052-3"          "GTSAMPLE1052-4"    

sample_names = c(
    "set1_macular_1",
    "set1_macular_2",
    "set1_macular_3",
    "set1_set1periphery_1",
    "set1_periphery_2",
    "set1_periphery_3",
    "set2_Macular_Punch_OS-3",
    "set2_Macular_Punch_OS-Unsorted-1",
    "set2_Macular_Punch_OS-Unsorted-2",
    "set2_Inferotemporal_OS-1",
    "set2_Inferotemporal_OS-2",
    "set2_Inferotemporal_OS-3",
    "set3_macular_1_fresh",
    "set3_macular_2_fresh",
    "set3_macular_3_fresh",
    "set3_macular_4_fresh",
    "set3_macular_1_prefrozen",
    "set3_macular_2_prefrozen",
    "set3_macular_3_prefrozen",
    "set3_macular_4_prefrozen"
)

name_mapping <- setNames(sample_names, sample)
seuratObj@meta.data$sample <- name_mapping[seuratObj@meta.data$sample]

```

# QC and selecting cells for further analysis

## Add %mitochondrial counts and cell cycle scores 
```{r, eval = FALSE}

## mitochondrial genes:
mt_genes <- c("ND1", "ND2", "COX1", "COX2", "ATP8", "ATP6", "COX3", "ND3", "ND4L", "ND4", "ND5", "ND6", "CYTB")

seuratObj[["percent.mt"]] <- PercentageFeatureSet(seuratObj, features = mt_genes)

s.genes <- cc.genes.updated.2019$s.genes
g2m.genes <- cc.genes.updated.2019$g2m.genes

seuratObj <- CellCycleScoring(seuratObj, s.features = s.genes, g2m.features = g2m.genes)
```

## QC
- Filter cells by:  
1, umi/nCount_RNA, low-quality cells (very low UMI counts) or doublets (very high UMI counts)  
2, gene/nFeature_RNA, remove potential low-quality cells and multiplets as above  
3, percent.mt, note that this step directly removes cells with high mt, the regress.out function later   
removes variations from mt and cell-cyle gene expression.   
4, our cutoff is chosen by visulizing data in Loupe browser    
5
== Load filtered Seurat obj instead of re-run the following steps 
```{r}
seuratObj <- qread("ocular_obj.qs")
# seuratObj <- readRDS("ocular_obj.rds")
```

- The object after pre-process
```{r}
seuratObj
table(seuratObj[[]]$sample)
head(seuratObj, 5)
```

### QC stats
```{r}
summary(seuratObj$nCount_RNA)
summary(seuratObj$nFeature_RNA)
summary(seuratObj$percent.mt)
```

UMI range is (`r min(seuratObj$nCount_RNA)`, `r max(seuratObj$nCount_RNA)`)
nfeatue range is (`r min(seuratObj$nFeature_RNA)`, `r max(seuratObj$nFeature_RNA)`)

Retinal ganglion cells are larger, results in a greater amount of total RNA per cell. 
Skip the nCount_RNA and nFeature_RNA filter as the gRed team did in their 2023 Cell Genomics paper.

Only remove cells with >5% mitochondrial expressions: `r length(seuratObj$percent.mt[seuratObj$percent.mt > 5])`

```{r, eval = FALSE}
seuratObj <- subset(seuratObj, subset = percent.mt < 5 &
    nFeature_RNA > 100 &
    nCount_RNA > 200)
```

### QC metrics as a violin plot 
- by initial clusters
```{r, fig.height=10, fig.width=12}
VlnPlot(seuratObj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), pt.size = 0, ncol = 3)
```

- by samples
```{r, fig.height=6, fig.width=12}
VlnPlot(seuratObj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
    group.by = 'sample',
    pt.size = 0.2, ncol = 3)
```


## SCTransform normalization, PCA and Clustering 
```{r, eval = FALSE}
# regressing out unwanted variation (mitochondrial, cell cyle, etc.)
seuratObj <- SCTransform(seuratObj, method = "glmGamPoi", vst.flavor = "v2",
        vars.to.regress = c("percent.mt", "S.Score", "G2M.Score"), verbose = TRUE)

seuratObj <- RunPCA(seuratObj, npcs = 40)
# ElbowPlot(seuratObj, ndims = 40)
seuratObj <- RunUMAP(seuratObj, reduction = "pca", dims = 1:30)
seuratObj <- FindNeighbors(seuratObj, reduction = "pca", dims = 1:30)
seuratObj <- FindClusters(seuratObj, resolution = 0.1)
table(seuratObj$seurat_clusters)
```

== Save processed object
```{r, eval = FALSE}
qsave(seuratObj, file = "ocular_obj.qs")
# saveRDS(seuratObj, file = "ocular_obj.rds")
```

## Visualize initial clusters
```{r, fig.height=7, fig.width=14}
p1 <- DimPlot(seuratObj, reduction = "umap", group.by = "sample") + theme(legend.position = "top")
p2 <- DimPlot(seuratObj, reduction = "umap", label = TRUE, repel = TRUE) + theme(legend.position = "top")
p1 + p2
```

```{r, fig.height=7, fig.width=7}
p1 <- DimPlot(seuratObj, reduction = "umap", group.by = "group") + theme(legend.position = "top")
p1
```

```{r, fig.height=10, fig.width=14}
DimPlot(seuratObj, reduction = "umap", label = TRUE, split.by = "sample", ncol = 4) + theme(legend.position = "top")
```

## Number of cells in each cluster
```{r}
plot_clusters <- function(srat) {
    count_table <- table(srat@meta.data$seurat_clusters, srat@meta.data$sample)
    count_mtx <- as.data.frame.matrix(count_table)
    count_mtx$clusters <- rownames(count_mtx)
    melt_mtx <- melt(count_mtx)

    celltype_size <- aggregate(value ~ clusters, data = melt_mtx, FUN = sum)
    celltype_size <- celltype_size %>% dplyr::arrange(value)
    celltype_size$clusters <- factor(celltype_size$clusters, levels = celltype_size$clusters)
    melt_mtx$clusters <- factor(melt_mtx$clusters, levels = celltype_size$clusters)
    colnames(melt_mtx)[2] <- "sample"

    p1 <- ggplot(celltype_size, aes(y = clusters, x = value)) +
    geom_bar(position = "dodge", stat = "identity", fill = "grey60") +
    theme_bw() + scale_x_log10() + xlab("Cells per cluster, log10 scale") + ylab("")

    p2 <- ggplot(melt_mtx, aes(x = clusters, y = value, fill = sample)) +
        geom_bar(position = "fill", stat = "identity") + theme_bw() + coord_flip() +
        scale_fill_manual(values=as.vector(distinctColorPalette(20))) +
        # scale_fill_brewer(palette = "Set2") +
        ylab("Fraction of cells in each sample") + xlab("cluster number") + theme(legend.position = "top")

    p2 + p1 + plot_layout(widths = c(3, 1))
}
```
```{r fig.height=8, fig.width=14}
plot_clusters(seuratObj)
```

# Data Integration  
To identify cell types present in all samples/conditions/modalities within our dataset,  
and would like to observe a representation of cells from both samples/conditions/modalities in every cluster.  
Reference: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/06_integration.md  
- Execute the normalization, variance estimation, and identification of the most variable genes for each sample   
- Perform integration of cells across conditions using the most variant genes to identify cells most similar to each other  

== Load object list `seuratObj.list`
```{r, eval = FALSE}
seuratObj.list <- qread("ocular_obj.list.qs")
# seuratObj.list <- readRDS("ocular_obj.list.rds")
```

== Load objects `ocular_obj.integrated`
```{r}
seuratObj.integrated <- qread("ocular_obj.integrated.qs")
# seuratObj.integrated <- readRDS("ocular_obj.integrated.rds")
table(seuratObj.integrated$seurat_clusters)
```

## Split object and SCTransform normalization  
split the dataset into a list of two seurat objects (GLA_KO and WT), 
normalize and identify variable features for each dataset independently.  
```{r, eval = FALSE}
seuratObj.list <- SplitObject(seuratObj, split.by = "sample")
seuratObj.list <- seuratObj.list[!grepl("set1", names(seuratObj.list))]
for (i in 1:length(seuratObj.list)) {
    seuratObj.list[[i]] <- SCTransform(seuratObj.list[[i]], method = "glmGamPoi", vst.flavor = "v2",
        vars.to.regress = c("percent.mt", "S.Score", "G2M.Score"), verbose = TRUE)
}
```

## Select features and prepare integration 
Select features that are repeatedly variable across datasets for integration, 
and run `PrepSCTIntegration`, which ensures that all necessary Pearson residuals have been calculated.  
```{r, eval = FALSE}
RGC_markers <- c('ENSMFAG00000036911','SYT2','SLC17A6','SNCG')
seuratObj.features <- SelectIntegrationFeatures(object.list = seuratObj.list, nfeatures = 3000)
seuratObj.features <- unique(c(seuratObj.features, RGC_markers))
seuratObj.list <- PrepSCTIntegration(object.list = seuratObj.list, anchor.features = seuratObj.features)
```

== Save object `seuratObj.list`
```{r, eval = FALSE}
qsave(seuratObj.list, file = "ocular_obj.list.qs")
# saveRDS(seuratObj.list, file = "ocular_obj.list.rds")
```

## Identify anchors and integrate the datasets
### using the samples with high RGCs as references (set3)
When using a set of specified references, anchors are first found between each query and each reference. 
The references are then integrated through pairwise integration. Each query is then mapped to the integrated reference.
Make sure to set normalization.method = **"SCT"**, and reduction = **"rpca"** for reciprocal PCA  
```{r, eval = FALSE}
anchors <- FindIntegrationAnchors(object.list = seuratObj.list, normalization.method = "SCT",
    reference = seq(7,14),
    anchor.features = seuratObj.features, reduction = "rpca")
seuratObj.integrated <- IntegrateData(anchorset = anchors, normalization.method = "SCT")
```


# Cluster on the integrated dataset
## PCA and Clustering  
```{r, eval = FALSE}
DefaultAssay(seuratObj.integrated) <- "integrated"
seuratObj.integrated <- RunPCA(seuratObj.integrated, npcs = 30)
# ElbowPlot(seuratObj.integrated, ndims = 30)

seuratObj.integrated <- RunUMAP(seuratObj.integrated, reduction = "pca", dims = 1:30)
seuratObj.integrated <- FindNeighbors(seuratObj.integrated, reduction = "pca", dims = 1:30)
# try 0.3-0.5, give 26, 28, 30 clusters
seuratObj.integrated <- FindClusters(seuratObj.integrated, resolution = 0.2)
table(seuratObj.integrated$seurat_clusters)
```

== Save integrated object
```{r, eval = FALSE}
qsave(seuratObj.integrated, file = "ocular_obj.integrated.qs")
# saveRDS(seuratObj.integrated, file = "ocular_obj.integrated.rds")
```

## Plots after integration   
```{r, fig.height=7, fig.width=14}
p1 <- DimPlot(seuratObj.integrated, reduction = "umap", group.by = "sample") + theme(legend.position = "top")
p2 <- DimPlot(seuratObj.integrated, reduction = "umap", label = TRUE, repel = TRUE) + NoLegend()
p1 + p2
```

```{r, fig.height=18, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = FALSE, split.by = "sample", ncol = 4) + theme(legend.position = "top")
```

```{r, fig.height=10, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = FALSE, split.by = "group", ncol = 2) + theme(legend.position = "top")
```

## Number of cells in each cluster
```{r}
plot_integrated_clusters <- function(srat) { 
    count_table <- table(srat@meta.data$seurat_clusters, srat@meta.data$sample)
    count_mtx <- as.data.frame.matrix(count_table)
    count_mtx$clusters <- rownames(count_mtx)
    melt_mtx <- melt(count_mtx)

    celltype_size <- aggregate(value ~ clusters, data = melt_mtx, FUN = sum)
    celltype_size <- celltype_size %>% dplyr::arrange(value)
    celltype_size$clusters <- factor(celltype_size$clusters, levels = celltype_size$clusters)
    melt_mtx$clusters <- factor(melt_mtx$clusters, levels = celltype_size$clusters)
    colnames(melt_mtx)[2] <- "sample"

    p1 <- ggplot(celltype_size, aes(y = clusters, x = value)) +
    geom_bar(position = "dodge", stat = "identity", fill = "grey60") +
    theme_bw() + scale_x_log10() + xlab("Cells per cluster, log10 scale") + ylab("")

    p2 <- ggplot(melt_mtx, aes(x = clusters, y = value, fill = sample)) +
        geom_bar(position = "fill", stat = "identity") + theme_bw() + coord_flip() +
        scale_fill_manual(values=as.vector(distinctColorPalette(20))) +
        ylab("Fraction of cells in each sample") + xlab("cluster number") + theme(legend.position = "top")

    p2 + p1 + plot_layout(widths = c(3, 1))
}
```
```{r fig.height=8, fig.width=14}
plot_integrated_clusters(seuratObj.integrated)
```


# Cell type annotation

## Apply marker based tool `sc-type`
```{r}
sctype <- function(obj) {
    # prepare marker genes from the database, category: Immune system, Liver, Pancreas, Kidney, Eye, Brain
    gs_list <- gene_sets_prepare("/mnt/data/apps/sc-type/ScTypeDB_short_NHPeye.xlsx", "NHP_Eye")

    # use a customized marker list
    # gs_list <- gene_sets_prepare("./ocular_customized_markers.xlsx", "Eye")
    # get cell-type by cell matrix
    # NOTE: scRNAseqData parameter should correspond to your input scRNA-seq matrix.
    # In case Seurat is used, it is either pbmc[["RNA"]]@scale.data (default),
    # pbmc[["SCT"]]@scale.data, in case sctransform is used for normalization,
    # or pbmc[["integrated"]]@scale.data, in case a joint analysis of multiple single-cell datasets is performed.
    es.max <- sctype_score(scRNAseqData = obj[['SCT']]@scale.data, scaled = TRUE,
                        gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)

    # merge by cluster
    metadata <- obj@meta.data
    clusters <- unique(obj@meta.data$seurat_clusters)
    cL_resutls <- do.call("rbind",
        lapply(
            clusters,
            function(cl) {
                es.max.cl <- sort(rowSums(es.max[, rownames(metadata[metadata$seurat_clusters == cl, ])]),
                    decreasing = TRUE)
                cell_scores <- data.frame(cluster = cl,
                        type = names(es.max.cl),
                        scores = es.max.cl,
                        ncells = sum(metadata$seurat_clusters == cl))
                head(cell_scores, 10)
            }
        )
    )
    sctype_scores <- cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)

    # set low-confident (low ScType score) clusters to "unknown"
    sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] <- "Unknown"
    print(data.frame(sctype_scores))

    # add cell type annotations to obj
    obj@meta.data$celltypes <- ""
    for (j in unique(sctype_scores$cluster)){
    cl_type <- sctype_scores[sctype_scores$cluster == j, ]
    obj@meta.data$celltypes[obj@meta.data$seurat_clusters == j] <-
        as.character(cl_type$type[1])
    }
    return(obj)
}
```

```{r}
seuratObj.integrated <- sctype(seuratObj.integrated)
```

## Show cell types on UMAP
```{r, fig.height=14, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = TRUE, repel = TRUE,
    group.by = 'celltypes') +
    theme(legend.position = "top")
table(seuratObj.integrated$celltypes)
```

```{r, fig.height=14, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = TRUE, repel = TRUE, split.by = 'sample', ncol = 4,
    group.by = 'celltypes') +
    theme(legend.position = "top")
```

```{r, fig.height=10, fig.width=14}
DimPlot(seuratObj.integrated, reduction = "umap", label = TRUE, repel = TRUE, split.by = 'group', ncol = 2,
    group.by = 'celltypes') +
    theme(legend.position = "top")
```

## Number of cells in each cell type
```{r}
plot_celltypes <- function(srat) {
    count_table <- table(srat@meta.data$celltypes, srat@meta.data$sample)
    count_mtx <- as.data.frame.matrix(count_table)
    count_mtx$celltypes <- rownames(count_mtx)
    melt_mtx <- melt(count_mtx)

    celltype_size <- aggregate(value ~ celltypes, data = melt_mtx, FUN = sum)
    celltype_size <- celltype_size %>% dplyr::arrange(value)
    celltype_size$celltypes <- factor(celltype_size$celltypes, levels = celltype_size$celltypes)
    melt_mtx$celltypes <- factor(melt_mtx$celltypes, levels = celltype_size$celltypes)
    colnames(melt_mtx)[2] <- "sample"

    p1 <- ggplot(celltype_size, aes(y = celltypes, x = value)) +
    geom_bar(position = "dodge", stat = "identity", fill = "grey60") +
    theme_bw() + scale_x_log10() + xlab("Cells per celltype, log10 scale") + ylab("")

    p2 <- ggplot(melt_mtx, aes(x = celltypes, y = value, fill = sample)) +
        geom_bar(position = "fill", stat = "identity") + theme_bw() + coord_flip() +
        scale_fill_manual(values=as.vector(distinctColorPalette(20))) +
        ylab("Fraction of cells in each sample") + xlab("celltype number") + theme(legend.position = "top")

    p2 + p1 + plot_layout(widths = c(3, 1))
}
```

```{r fig.height=6, fig.width=14}
plot_celltypes(seuratObj.integrated)
```



# Cluster markers analysis
== Load all cluster markers obj
```{r}
all.markers <- readRDS("all.cluster_markers.rds")
head(all.markers)
```

## Find all markers
```{r, eval = FALSE}
seuratObj.integrated <- PrepSCTFindMarkers(seuratObj.integrated, assay = "SCT", verbose = TRUE)
plan("multicore", workers = 8)
all.markers <- FindAllMarkers(seuratObj.integrated, assay = "SCT", verbose = TRUE)
plan(sequential)
```

## Save marker genes
```{r, eval = FALSE}
all.markers <- all.markers %>% relocate(gene)
write_tsv(all.markers, "all.cluster_markers.tsv")
top20 <- all.markers %>%
    group_by(cluster) %>%
    top_n(n = 25, wt = avg_log2FC)
write_tsv(top20, "top25.cluster_markers.tsv")
saveRDS(all.markers, file = "all.cluster_markers.rds")
```

## Heatmap with top genes {.tabset}

### Markers for each Cluster
- showing expression in 10% random cells for each cluster  
```{r, fig.height=20, fig.width=12}
Idents(seuratObj.integrated) <- "seurat_clusters"
DefaultAssay(seuratObj.integrated) <- "SCT"
top10 <- all.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC)

DoHeatmap(subset(seuratObj.integrated, downsample = 100), features = top10$gene, size = 3) +
    theme(legend.position = 'none', plot.margin = margin(t = 20))
```

# RETINAL GANGLION CELLS (RGCs) ==

```{r}
sn <- seuratObj.integrated
DefaultAssay(sn) <- "RNA"
table(Idents(sn))

# hgd()
# hgd_view()

# RGCs are the projection neurons of the retina
# markers of RGCs are:

# RBPMS, which is not correctly annotated in macaque reference genome 
# transcript ID for this is "ENSMFAG00000036911"

VlnPlot(sn, features = "ENSMFAG00000036911")

# glutamate transporter VGLUT2 (SLC17A6 gene)
VlnPlot(sn, features = "SLC17A6")

# gamma synuclein (SNCG)
VlnPlot(sn, features = "SNCG")

# synaptotagmin 2 (SYT2)
VlnPlot(sn, features = "SYT2")

#### markers from Kamil
VlnPlot(sn, features = "THY1")
VlnPlot(sn, features = "NEFL")
VlnPlot(sn, features = "RBPMS2")
VlnPlot(sn, features = "OPN4")
VlnPlot(sn, features = "POU4F1")
VlnPlot(sn, features = "POU4F2")

# new markers
VlnPlot(sn, features = "ELAVL2")
VlnPlot(sn, features = "RBFOX3")
VlnPlot(sn, features = "IRX1")
VlnPlot(sn, features = "IRX2")
VlnPlot(sn, features = "NEFH")
```


## plot % RGCs for each sample/tissue/source
```{r}
Idents(sn) <- "celltypes"
cell_type_to_plot <- "Retinal ganglion cells"
percentage_data <- sn@meta.data %>%
    group_by(sample) %>%
    summarise(Percentage = sum(celltypes == cell_type_to_plot) / n()) %>%
    ungroup()

# pdf("Percent_RGC_per_sample.pdf", 8,6)
ggplot(percentage_data, aes(x = sample, y = Percentage, fill = sample)) +
    geom_bar(stat = "identity") +
    labs(title = paste("Percentage of", cell_type_to_plot), fill = "sample") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1), plot.title = element_text(hjust = 0.5)) +
    scale_y_continuous(labels = scales::percent) +
    theme(plot.margin = margin(5, 5, 5, 20, "mm")) 
# dev.off()
```

## RGC SUBTYPES 
```{r}
RGCs <- subset(sn, idents = cell_type_to_plot)
RGCs
# re-running SCTransform normalization, PCA and Clustering           
RGCs <- SCTransform(RGCs, method = "glmGamPoi", vst.flavor = "v2", verbose = TRUE)
RGCs <- RunPCA(RGCs, npcs = 40)
# ElbowPlot(RGCs)
RGCs <- RunUMAP(RGCs, reduction = "pca", dims = 1:15)
RGCs <- FindNeighbors(RGCs, reduction = "pca", dims = 1:15)
RGCs <- FindClusters(RGCs, resolution = 0.1)
table(RGCs$seurat_clusters)
```

### Visualize clusters
```{r, fig.height=7, fig.width=7}

DimPlot(RGCs, reduction = "umap", label = TRUE, repel = TRUE) + theme(legend.position = "top")
```

```{r, fig.height=7, fig.width=7}
p1 <- DimPlot(RGCs, reduction = "umap", group.by = "group") + theme(legend.position = "top")
p1
```

```{r, fig.height=10, fig.width=14}
DimPlot(RGCs, reduction = "umap", label = TRUE, split.by = "sample", ncol = 4) + theme(legend.position = "top")
```

## Redo integration for RGCs only

```{r}
RGC.integrated <- qread("RGC.integrated.qs")
table(RGC.integrated$seurat_clusters)
```

### load previous single dataset objects and do cell type annotation
The data has been already normalized and clustered, but no cell type annotation  
```{r, eval = FALSE}
macula_sort_unsort_set2 <- qread('../Ocular_snRNAseq_macula_periphery_Wunsorted/ocular_obj.qs')
macula_comp_dissection_set3 <- qread('../Ocular_snRNAseq_macula_comp_dissections/ocular_obj.qs')
```

```{r, eval = FALSE}
macula_sort_unsort_set2 <- sctype(macula_sort_unsort_set2)
macula_comp_dissection_set3 <- sctype(macula_comp_dissection_set3)
```

### extract RGCs
```{r, eval = FALSE}
Idents(macula_sort_unsort_set2) <- "celltypes"
set2_RGCs <- subset(macula_sort_unsort_set2, idents = "Retinal ganglion cells")
set2_RGCs$sample <- gsub('FR2749M_', 'set2_', set2_RGCs$sample)

Idents(macula_comp_dissection_set3) <- "celltypes"
set3_RGCs <- subset(macula_comp_dissection_set3, idents = "Retinal ganglion cells")
sample = unique(set3_RGCs$sample)   

sample_names = c(
    "set3_macular_1_fresh",
    "set3_macular_2_fresh",
    "set3_macular_3_fresh",
    "set3_macular_4_fresh",
    "set3_macular_1_prefrozen",
    "set3_macular_2_prefrozen",
    "set3_macular_3_prefrozen",
    "set3_macular_4_prefrozen"
)

name_mapping <- setNames(sample_names, sample)
set3_RGCs@meta.data$sample <- name_mapping[set3_RGCs@meta.data$sample]
```

### Integrate RGCs from set2 and set3
```{r, eval = FALSE}
## Split object and SCTransform normalization  
RGC.list <- c(SplitObject(set2_RGCs, split.by = "sample"), SplitObject(set3_RGCs, split.by = "sample"))
RGC.list <- Filter(function(i) { ncol(i) >= 50 }, RGC.list)

for (i in 1:length(RGC.list)) {
    RGC.list[[i]] <- SCTransform(RGC.list[[i]], method = "glmGamPoi", vst.flavor = "v2",
        vars.to.regress = c("percent.mt", "S.Score", "G2M.Score"), verbose = TRUE) %>%
    RunPCA(npcs = 30)
}

## Select features and prepare integration 
RGC_markers <- c('ENSMFAG00000036911','SYT2','SLC17A6','SNCG')
RGC.features <- SelectIntegrationFeatures(object.list = RGC.list, nfeatures = 3000)
RGC.features <- unique(c(RGC.features, RGC_markers))
RGC.list <- PrepSCTIntegration(object.list = RGC.list, anchor.features = RGC.features)

## Identify anchors and integrate the datasets
### using the samples with high RGCs as references (set3)
anchors <- FindIntegrationAnchors(object.list = RGC.list, normalization.method = "SCT",
    reference = seq(4,11),
    anchor.features = RGC.features, reduction = "rpca")
RGC.integrated <- IntegrateData(anchorset = anchors, k.weight = 70, normalization.method = "SCT") # one sample has only 73 cells, k.weight should less than that

# add groups
group <- ifelse(grepl("set2", RGC.integrated$sample), "macula_sort_unsort_set2", "macula_comp_dissection_set3")
RGC.integrated  <- AddMetaData(RGC.integrated, metadata = group, col.name = "group")
```

== Save object RGC.list
```{r, eval = FALSE}
qsave(RGC.list, file = "RGC.list.qs")
```

### sub-cluster on the integrated RGCs
```{r, eval = FALSE}
DefaultAssay(RGC.integrated) <- "integrated"
RGC.integrated <- RunPCA(RGC.integrated, npcs = 30)
# ElbowPlot(RGC.integrated, ndims = 30)

RGC.integrated <- RunUMAP(RGC.integrated, reduction = "pca", dims = 1:20)
RGC.integrated <- FindNeighbors(RGC.integrated, reduction = "pca", dims = 1:20)
# try 0.3-0.5, give 26, 28, 30 clusters
RGC.integrated <- FindClusters(RGC.integrated, resolution = 0.1)
table(RGC.integrated$seurat_clusters)
```

== Save integrated object
```{r, eval = FALSE}
qsave(RGC.integrated, file = "RGC.integrated.qs")
```

### Plots after integration   
```{r, fig.height=7, fig.width=14}
p1 <- DimPlot(RGC.integrated, reduction = "umap", group.by = "sample") + theme(legend.position = "top")
p2 <- DimPlot(RGC.integrated, reduction = "umap", label = TRUE, repel = TRUE) + NoLegend()
p1 + p2
```

```{r, fig.height=18, fig.width=14}
DimPlot(RGC.integrated, reduction = "umap", label = FALSE, split.by = "sample", ncol = 4) + theme(legend.position = "top")
```

```{r, fig.height=10, fig.width=14}
DimPlot(RGC.integrated, reduction = "umap", label = TRUE, split.by = "group", ncol = 2) + theme(legend.position = "top")
```

** plot markers **
```{r}
DefaultAssay(RGC.integrated) <- "RNA"

# RBPMS, which is not correctly annotated in macaque reference genome 
# transcript ID for this is "ENSMFAG00000036911"
VlnPlot(RGC.integrated, features = "ENSMFAG00000036911")

# glutamate transporter VGLUT2 (SLC17A6 gene)
VlnPlot(RGC.integrated, features = "SLC17A6")

# gamma synuclein (SNCG)
VlnPlot(RGC.integrated, features = "SNCG")

# synaptotagmin 2 (SYT2)
VlnPlot(RGC.integrated, features = "SYT2")

#### markers from Kamil
VlnPlot(RGC.integrated, features = "THY1")
VlnPlot(RGC.integrated, features = "NEFL")
VlnPlot(RGC.integrated, features = "RBPMS2")
VlnPlot(RGC.integrated, features = "OPN4")
VlnPlot(RGC.integrated, features = "POU4F1")
VlnPlot(RGC.integrated, features = "POU4F2")

# new markers
VlnPlot(RGC.integrated, features = "ELAVL2")
VlnPlot(RGC.integrated, features = "RBFOX3")
VlnPlot(RGC.integrated, features = "IRX1")
VlnPlot(RGC.integrated, features = "IRX2")
VlnPlot(RGC.integrated, features = "NEFH")
```

```{r}

RGC.integrated <- PrepSCTFindMarkers(RGC.integrated, assay = "SCT", verbose = TRUE)
RGC.markers <- FindAllMarkers(RGC.integrated, assay = "SCT", verbose = TRUE)
# 25 top markers of clusters
RGC.markers %>%
    group_by(cluster) %>%
    top_n(n = 25, wt = avg_log2FC) -> top25

# checking markers of RGC subtypes

# Midget RGCs
VlnPlot(RGC.integrated, features = "IL1RAPL2")

# Midget ON cells
VlnPlot(RGC.integrated, features = "EOMES")
VlnPlot(RGC.integrated, features = "IRX3")

# OFF cells
VlnPlot(RGC.integrated, features = "MEIS2")

### clusters 0, 2, 6 are OFF Midget RGCs whereas clusters 1 is ON Midget RGCs

# Parasol RGCs
VlnPlot(RGC.integrated, features = "RBPMS2")
VlnPlot(RGC.integrated, features = "SPP1")

### clusters 3,4,5 are parasol cells

# OFF parasol RGCs
VlnPlot(RGC.integrated, features = "PROX1")
VlnPlot(RGC.integrated, features = "CA8")

### cluster 3 contains OFF parasol RGCs

# ON parasol RGCs
VlnPlot(RGC.integrated, features = "CHRNA2")

### cluster 4 is ON parasol RGCs

# rare RGC subtypes, cluster 5
VlnPlot(RGC.integrated, features = "FOXP2")
VlnPlot(RGC.integrated, features = "BNC2")
VlnPlot(RGC.integrated, features = "SATB2")

```

** plot subtypes **
```{r}
DefaultAssay(RGC.integrated) <- "integrated"
cluster_to_celltype <- c('0' = 'Midget OFF', '2' = 'Midget OFF', '6' = 'Midget OFF',
    '1' = 'Midget ON', '3' = 'Parasol OFF', '4' = 'Parasol ON', '5' = 'Other')

RGC.integrated@meta.data$subtype <- cluster_to_celltype[RGC.integrated@meta.data$seurat_clusters]
table(RGC.integrated$subtype)

# Calculate the percentage of each cell type
celltype_percentages <- table(RGC.integrated$subtype) / length(RGC.integrated$subtype) * 100

# Convert to dataframe for ggplot
celltype_df <- as.data.frame(celltype_percentages)
names(celltype_df) <- c("subtype", "percentage")

# Plot

# Manually specify colors for each cell type
manual_colors <- c("Midget OFF" = "#00bfff", 
                   "Midget ON" = "#0073ff", 
                   "Parasol OFF" = "#ff8c00", 
                   "Parasol ON" = "#ff5500", 
                   "Other" = "purple")
                   
ggplot(celltype_df, aes(x = "", y = percentage, fill = subtype)) +
    geom_bar(width = 1, stat = "identity") +
    coord_polar("y", start = 0) +
    theme_void() +
    labs(fill = "Cell Type") +
    geom_text(aes(label = paste0(round(percentage, 1), "%")), 
                position = position_stack(vjust = 0.5)) +
    scale_fill_manual(values = manual_colors) +
    ggtitle("Percentage of RGC sub-types")

```